src/Sylius/Bundle/CoreBundle/OrderProcessing/OrderTaxesApplicator.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\OrderProcessing;
+
+use Sylius\Bundle\CoreBundle\Distributor\TaxesDistributorInterface;
+use Sylius\Bundle\CoreBundle\Provider\DefaultTaxZoneProviderInterface;
+use Sylius\Component\Addressing\Matcher\ZoneMatcherInterface;
+use Sylius\Component\Addressing\Model\AddressInterface;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\Taxation\Calculator\CalculatorInterface;
+use Sylius\Component\Taxation\Resolver\TaxRateResolverInterface;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class OrderTaxesApplicator implements OrderTaxesApplicatorInterface
+{
+    /**
+     * @var CalculatorInterface
+     */
+    protected $calculator;
+
+    /**
+     * @var DefaultTaxZoneProviderInterface
+     */
+    protected $defaultTaxZoneProvider;
+
+    /**
+     * @var FactoryInterface
+     */
+    protected $adjustmentFactory;
+
+    /**
+     * @var TaxesDistributorInterface
+     */
+    protected $orderUnitTaxesDistributor;
+
+    /**
+     * @var TaxRateResolverInterface
+     */
+    protected $taxRateResolver;
+
+    /**
+     * @var ZoneMatcherInterface
+     */
+    protected $zoneMatcher;
+
+    /**
+     * @param CalculatorInterface $calculator
+     * @param DefaultTaxZoneProviderInterface $defaultTaxZoneProvider
+     * @param FactoryInterface $adjustmentFactory
+     * @param TaxesDistributorInterface $orderUnitTaxesDistributor
+     * @param TaxRateResolverInterface $taxRateResolver
+     * @param ZoneMatcherInterface $zoneMatcher
+     */
+    public function __construct(
+        CalculatorInterface $calculator,
+        DefaultTaxZoneProviderInterface $defaultTaxZoneProvider,
+        FactoryInterface $adjustmentFactory,
+        TaxesDistributorInterface $orderUnitTaxesDistributor,
+        TaxRateResolverInterface $taxRateResolver,
+        ZoneMatcherInterface $zoneMatcher
+    ) {
+        $this->calculator = $calculator;
+        $this->defaultTaxZoneProvider = $defaultTaxZoneProvider;
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->orderUnitTaxesDistributor = $orderUnitTaxesDistributor;
+        $this->taxRateResolver = $taxRateResolver;
+        $this->zoneMatcher = $zoneMatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order)
+    {
+        if ($order->getItems()->isEmpty()) {
+            return;
+        }
+
+        $this->clearAdjustments($order);
```

Comment:
This should be called on empty cart also, so we won't have a tax adjustments applied to an empty order.

I would rename the method to `clearTaxes`.